src/Xamarin.Android.Build.Tasks/Tasks/BuildApk.cs

```diff
@@ -544,25 +549,119 @@ void AddRuntimeLibraries (ZipArchiveEx apk, string [] supportedAbis)
+			}
+
+			return String.Compare (Path.GetFileName (link), "wrap.sh", StringComparison.Ordinal) == 0;
+		}
+
+		bool IncludeNativeLibrary (ITaskItem item)
+		{
+			if (IncludeWrapSh)
+				return true;
+
+			return !IsWrapperScript (item.ItemSpec, item.GetMetadata ("Link"));
+		}
+
+		string GetArchiveFileName (ITaskItem item)
+		{
+			string archiveFileName = item.GetMetadata ("ArchiveFileName");
+			if (!String.IsNullOrEmpty (archiveFileName))
+				return archiveFileName;
+
+			if (!IsWrapperScript (item.ItemSpec, item.GetMetadata ("Link"))) {
+				return null;
+			}
+
+			return "wrap.sh";
+		}
+
 		private void AddNativeLibraries (ArchiveFileList files, string [] supportedAbis)
 		{
 			var frameworkLibs = FrameworkNativeLibraries.Select (v => new LibInfo {
 				Path = v.ItemSpec,
+				Link = v.GetMetadata ("Link"),
 				Abi = GetNativeLibraryAbi (v),
-				ArchiveFileName = v.GetMetadata ("ArchiveFileName")
+				ArchiveFileName = GetArchiveFileName (v)
 			});
 
 			AddNativeLibraries (files, supportedAbis, frameworkLibs);
 
 			var libs = NativeLibraries.Concat (BundleNativeLibraries ?? Enumerable.Empty<ITaskItem> ())
+				.Where (v => IncludeNativeLibrary (v))
 				.Select (v => new LibInfo {
 						Path = v.ItemSpec,
+						Link = v.GetMetadata ("Link"),
 						Abi = GetNativeLibraryAbi (v),
-						ArchiveFileName = v.GetMetadata ("ArchiveFileName")
+						ArchiveFileName = GetArchiveFileName (v)
 					}
 				);
 
 			AddNativeLibraries (files, supportedAbis, libs);
+
+			if (String.IsNullOrWhiteSpace (CheckedBuild))
+				return;
+
+			string mode = CheckedBuild;
+			string sanitizerName;
+			if (String.Compare ("asan", mode, StringComparison.Ordinal) == 0) {
+				sanitizerName = "asan";
+			} else if (String.Compare ("ubsan", mode, StringComparison.Ordinal) == 0) {
+				sanitizerName = "ubsan_standalone";
+			} else {
+				Log.LogWarning ($"Unknown checked build mode '{CheckedBuild}'");
+				return;
+			}
+
+			if (!IncludeWrapSh) {
+				Log.LogError ("Checked builds require the wrapper script to be packaged. Please set the `$(AndroidIncludeWrapSh) MSBuild property to `true` in your project");
+				return;
+			}
+
+			if (!libs.Any (info => IsWrapperScript (info.Path, info.Link))) {
+				// TODO: consider including various wrapper scripts with XA and adding them here
+				Log.LogError ($"Checked builds require the wrapper script to be packaged. Please add `wrap.sh` appropriate for the {CheckedBuild} checker to your project");
+				return;
+			}
+
+			NdkUtil.Init (AndroidNdkDirectory);
+			string clangDir = NdkUtil.GetClangDeviceLibraryPath (AndroidNdkDirectory);
+			if (String.IsNullOrEmpty (clangDir)) {
+				Log.LogError ($"Unable to find the clang compiler directory. Is NDK installed?");
+				return;
+			}
+
+			foreach (string abi in supportedAbis) {
+				string clangAbi = MonoAndroidHelper.MapAndroidAbiToClang (abi);
+				if (String.IsNullOrEmpty (clangAbi)) {
+					Log.LogError ($"Unable to map Android ABI {abi} to clang ABI");
```

Comment:
`LogError()` should be `LogCodedError()`, along with corresponding docs in `Documentation/guides/messages`.

Reply:
This message will be logged only for checked builds, that is not on the end user machine.